src/hotspot/share/opto/stringopts.cpp

```diff
@@ -277,6 +277,19 @@ void StringConcat::eliminate_unneeded_control() {
       C->gvn_replace_by(n, n->in(0)->in(0));
       // get rid of the other projection
       C->gvn_replace_by(n->in(0)->as_If()->proj_out(false), C->top());
+    } else if (n->is_Region()) {
+      Node* iff = n->in(1)->in(0);
+      assert(iff->is_If(), "no if for the diamond");
```

Comment:
I assume that the only Region node listed in _control list is diamond Region.
May be add assert to check that it is diamond region. Up to you.
I would like to have separate RFE to have `is_diamond_region()` - we seems check it in few places.

Reply:
I pushed an update. Is this what you had in mind?

Reply:
yes